safenode/src/network/event.rs

```diff
@@ -0,0 +1,172 @@
+                    warn!("RequestResponseError: {e:?}");
+                }
+            }
+            // handle Kademlia events
+            SwarmEvent::Behaviour(SafeNodeEvent::Kademlia(event)) => match event {
+                KademliaEvent::OutboundQueryProgressed {
+                    id,
+                    result: QueryResult::StartProviding(_),
+                    ..
+                } => {
+                    let sender: oneshot::Sender<Result<()>> = self
+                        .pending_start_providing
+                        .remove(&id)
+                        .ok_or(Error::Other(
+                            "Completed query to be previously pending.".to_string(),
+                        ))?;
+                    let _ = sender.send(Ok(()));
+                }
+                KademliaEvent::OutboundQueryProgressed {
+                    id,
+                    result:
+                        QueryResult::GetProviders(Ok(GetProvidersOk::FoundProviders {
+                            providers, ..
+                        })),
+                    ..
+                } => {
+                    if let Some(sender) = self.pending_get_providers.remove(&id) {
+                        sender
+                            .send(providers)
+                            .map_err(|_| Error::Other("Receiver not to be dropped".to_string()))?;
+
+                        // Finish the query. We are only interested in the first result.
+                        self.swarm
+                            .behaviour_mut()
+                            .kademlia
+                            .query_mut(&id)
+                            .ok_or(Error::Other("Query should be exist".to_string()))?
+                            .finish();
+                    }
+                }
+                _ => {}
+            },
+            SwarmEvent::Behaviour(SafeNodeEvent::Mdns(mdns_event)) => match *mdns_event {
+                mdns::Event::Discovered(list) => {
+                    for (peer_id, multiaddr) in list {
+                        info!("Node discovered: {multiaddr:?}");
+                        let _routing_update = self
+                            .swarm
+                            .behaviour_mut()
+                            .kademlia
+                            .add_address(&peer_id, multiaddr);
+                    }
+                    self.event_sender.send(NetworkEvent::PeerDiscoverd).await?;
+                }
+                mdns::Event::Expired(_) => {
+                    info!("mdns peer expired");
+                }
+            },
+            SwarmEvent::NewListenAddr { address, .. } => {
+                let local_peer_id = *self.swarm.local_peer_id();
+                info!(
+                    "Local node is listening on {:?}",
+                    address.with(Protocol::P2p(local_peer_id.into()))
+                );
+            }
+            SwarmEvent::IncomingConnection { .. } => {}
```

Comment:
Just a note: I think this is something we could use for the bayesian function for suspicious behaviour detection, as David talked about.